cli/azd/internal/repository/app_init.go

```diff
@@ -467,3 +495,179 @@ func (i *Initializer) prjConfigFromDetect(
 
 	return config, nil
 }
+
+type property struct {
+	key   string
+	value string
+}
+type propertyMergeFunc func(string, string) string
+
+const azureProfileName = "azure"
+
+var placeholderPostgresJdbcUrl = "jdbc:postgresql://" + internal.ToEnvPlaceHolder(internal.EnvNamePostgresHost) +
+	":" + internal.ToEnvPlaceHolder(internal.EnvNamePostgresPort) +
+	"/" + internal.ToEnvPlaceHolder(internal.EnvNamePostgresDatabase)
+
+var postgresqlProperties = []property{
+	{"spring.datasource.url", placeholderPostgresJdbcUrl},
+	{"spring.datasource.username", internal.ToEnvPlaceHolder(internal.EnvNamePostgresUsername)},
+	{"spring.datasource.password", internal.ToEnvPlaceHolder(internal.EnvNamePostgresPassword)},
+}
+
+var applicationPropertiesRelativePath = filepath.Join("src", "main", "resources",
+	"application.properties")
+var applicationAzurePropertiesRelativePath = filepath.Join("src", "main", "resources",
+	"application-"+azureProfileName+".properties")
+
+// todo: support other file suffix. Example: application.yml, application.yaml
+func addPostgresqlConnectionProperties(projectPath string) error {
+	err := addPostgresqlConnectionPropertiesIntoPropertyFile(projectPath)
+	if err != nil {
+		return err
+	}
+	return activeAzureProfile(projectPath)
+}
+
+func addPostgresqlConnectionPropertiesIntoPropertyFile(projectPath string) error {
+	filePath := filepath.Join(projectPath, applicationAzurePropertiesRelativePath)
+	return updatePropertyFile(filePath, postgresqlProperties, keepNewValue)
+}
+
+func keepNewValue(_ string, newValue string) string {
+	return newValue
+}
+
+func activeAzureProfile(projectPath string) error {
+	filePath := filepath.Join(projectPath, applicationPropertiesRelativePath)
+	var newProperties = []property{
+		{"spring.profiles.active", azureProfileName},
+	}
+	return updatePropertyFile(filePath, newProperties, appendToCommaSeparatedValues)
+}
+
+func appendToCommaSeparatedValues(commaSeparatedValues string, newValue string) string {
+	if commaSeparatedValues == "" {
+		return newValue
+	}
+	var values []string
+	for _, value := range strings.SplitN(commaSeparatedValues, ",", -1) {
+		value = strings.TrimSpace(value)
+		if value != "" {
+			values = append(values, value)
+		}
+	}
+	if !contains(values, azureProfileName) {
+		values = append(values, azureProfileName)
+	}
+	return strings.Join(values, ",")
+}
+
+func contains(a []string, x string) bool {
+	for _, n := range a {
+		if x == n {
+			return true
+		}
+	}
+	return false
+}
+
+func updatePropertyFile(filePath string, newProperties []property, function propertyMergeFunc) error {
+	err := createFileIfNotExist(filePath)
```

Comment:
Would it be simpler to just have a single function that opens the file, and just does a one-pass update? 

I know that we've created a lot of functions here to attempt to break down the complexity. In doing so, they all "look right" individually, but if we zoom out to see the whole algorithm: 

- If `createFileNotExists` succeeds, but we fail in `readProperties`, we create an empty file
- We also create the file again in `writeProperties`

FWIW, this is related to what we had talked about creating too many functions.